### PR TITLE
Moq1100 incorrectly firing on nullable parameters

### PR DESCRIPTION
**Pull request by rjmurillo**
Update type comparison to use `ClassifyConversion` to check if the types are assignable, rather than stringly strictly equal. 

## Changes:

- Add check for types based on assignability (which also covers user-defined implicit and explicit conversions), rather than strictly equal with types. 
- Add logging for test type to aid in debugging cases
- Update detection (span for diagnostic is tighter now)
- Add repro for issue reported in #172 (thanks @marcovr!)
- Add test cases for user-defined implicit and explicit conversions (thanks @MattKotsenas!)

### CODE REVIEW

File: src/Moq.Analyzers/CallbackSignatureShouldMatchMockedMethodAnalyzer.cs [hunk 1]
```diff
@@ -58,31 +58,52 @@ private static void Analyze(SyntaxNodeAnalysisContext context)
...
+            // Check if there's an implicit conversion or identity conversion
+            Conversion conversion = context.SemanticModel.Compilation.ClassifyConversion(lambdaParameterTypeSymbol, mockedMethodTypeSymbol);
 
-                if (!string.Equals(mockedMethodTypeName, lambdaParameterTypeName, StringComparison.Ordinal))
-                {
-                    Diagnostic diagnostic = callbackLambda.ParameterList.CreateDiagnostic(Rule);
-                    context.ReportDiagnostic(diagnostic);
-                }
+            // Where a conversion exists, even if it is explicit
+            if (conversion.Exists && conversion is not { IsImplicit: false, IsIdentity: false, IsExplicit: false })
```
Comment (MattKotsenas) [inline src/Moq.Analyzers/CallbackSignatureShouldMatchMockedMethodAnalyzer.cs:100]: I don't understand this part (and may be my own misunderstanding).

If there's an implicit or identity conversion, then continue, but if there's an explicit conversion available, then don't you also need to check if the cast is being performed here?

Comment (MattKotsenas) [reply src/Moq.Analyzers/CallbackSignatureShouldMatchMockedMethodAnalyzer.cs:100]: Also, this code doesn't match the comment on line 96.

File: tests/Moq.Analyzers.Test/CallbackSignatureShouldMatchMockedMethodCodeFixTests.cs [hunk 2]
```diff
@@ -91,6 +119,10 @@ internal interface IFoo
                 int Do(int i, string s, DateTime dt);
 
                 int Do(List<string> l);
+
+                bool Do(object? bar);
+
+                // bool Do(long bar);
```
Comment (MattKotsenas) [inline tests/Moq.Analyzers.Test/CallbackSignatureShouldMatchMockedMethodCodeFixTests.cs:125]: nit: comment

Comment (rjmurillo) [reply tests/Moq.Analyzers.Test/CallbackSignatureShouldMatchMockedMethodCodeFixTests.cs]: Fixed

File: src/Moq.Analyzers/CallbackSignatureShouldMatchMockedMethodAnalyzer.cs [hunk 3]
```diff
@@ -58,31 +58,101 @@ private static void Analyze(SyntaxNodeAnalysisContext context)
...
+                return true;
 
-                if (!string.Equals(mockedMethodTypeName, lambdaParameterTypeName, StringComparison.Ordinal))
+            // Fall back to seeking an implicit conversion operator
+            // The Conversion type will not show a user defined implicit conversion, so we have to seek it out explicitly
+            case false:
                 {
-                    Diagnostic diagnostic = callbackLambda.ParameterList.CreateDiagnostic(Rule);
-                    context.ReportDiagnostic(diagnostic);
+                    foreach (ISymbol? member in destination.GetMembers())
```
Comment (MattKotsenas) [inline src/Moq.Analyzers/CallbackSignatureShouldMatchMockedMethodAnalyzer.cs:136]: This doesn't feel right to me. `Conversion` has an `IsUserDefined` method on it, why doesn't that work?

Comment (rjmurillo) [reply src/Moq.Analyzers/CallbackSignatureShouldMatchMockedMethodAnalyzer.cs]: Not sure why that doesn't work. Given the following scenario

```csharp
new Mock<IFoo>().Setup(x => x.Do((CustomType)"42")).Returns((CustomType i) => true);

internal interface IFoo
{
    bool Do(CustomType custom);
}

public class CustomType
{
    public int Value { get; }

    public CustomType(int value)
    {
        Value = value;
    }

    // User-defined conversions
    public static implicit operator CustomType(int value)
    {
        return new CustomType(value);
    }

    public static explicit operator CustomType(string str)
    {
        return new CustomType(int.Parse(str));
    }
}
```

The property for `conversion.IsUserDefined` is false. The only things that are true is `Exists`, `IsIdentity` and `IsImplicit`.

In the case `new Mock<IFoo>().Setup(x => x.Do(42)).Returns((CustomType i) => true);` `conversion` has `Explicit`, `IsUserDefined` and `IsImplicit` true.

**Review by MattKotsenas — APPROVED** [review]

**Pull request merged**